service/src/java/org/apache/hive/service/servlet/OTELExporter.java

```diff
@@ -0,0 +1,225 @@
+        }
+        initSpan.end();
+
+        for (QueryDisplay.TaskDisplay task : lQuery.getQueryDisplay().getTaskDisplays()) {
+          if (task.getReturnValue() != null && task.getEndTime() != null) {
+            completedTasks.add(task.getTaskId());
+            parentContext = Context.current().with(rootspan);
+            tracer.spanBuilder(queryID + " - " + task.getTaskId() + " - live")
+                    .setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                    .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan()
+                    .end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+        queryIdToSpanMap.put(queryID,rootspan);
+        queryIdToTasksMap.put(queryID,completedTasks);
+      }
+    }
+
+    List<String> historicalQueryIDs = new ArrayList<>();
+    for (QueryInfo hQuery : historicalQueries) {
+      String hQueryId = hQuery.getQueryDisplay().getQueryId();
+      historicalQueryIDs.add(hQueryId);
+      Span rootspan = queryIdToSpanMap.get(hQueryId);
+
+      //For queries that were live till last loop but have ended before start of this loop
+      if (rootspan != null) {
+        for (QueryDisplay.TaskDisplay task : hQuery.getQueryDisplay().getTaskDisplays()) {
+          if (!queryIdToTasksMap.get(hQueryId).contains(task.getTaskId())) {
+            queryIdToTasksMap.get(hQueryId).add(task.getTaskId());
+            Context parentContext = Context.current().with(rootspan);
+            tracer.spanBuilder(hQueryId+ " - " + task.getTaskId() + " - completed")
+                    .setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                    .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan()
+                    .end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+
+        //Update the rootSpan name & attributes before ending it
+        rootspan.updateName(hQueryId + " - completed").setAllAttributes(addQueryAttributes(hQuery)).end();
+        queryIdToSpanMap.remove(hQueryId);
+        queryIdToTasksMap.remove(hQueryId);
+
+        historicalQueryId.add(hQueryId);
+      }
+
+      //For queries that already ended either before OTEL service started or in between OTEL loops
+      if (historicalQueryId.add(hQuery.getQueryDisplay().getQueryId())) {
+        rootspan = tracer.spanBuilder(hQuery.getQueryDisplay().getQueryId() + " - completed")
+                .startSpan();
+        Context parentContext = Context.current().with(rootspan);
+
+        Span initSpan = tracer.spanBuilder(hQuery.getQueryDisplay().getQueryId() + " - completed").setParent(parentContext).startSpan()
+                .setAttribute("queryID", hQuery.getQueryDisplay().getQueryId())
+                .setAttribute("queryString", hQuery.getQueryDisplay().getQueryString())
+                .setAttribute("Begin Time", hQuery.getBeginTime());
```

Comment:
shouldn't we instead call ``addQueryAttributes`` here?

Reply:
It would make initSpan and rootSpan storing the same content in case of historical queries. Would be better if we keep only these minimal attributes in initSpan so that there's consistency between live and historical traces.